lstchain/mc/sensitivity.py

```diff
@@ -387,31 +411,46 @@ def find_best_cuts_sens(simtelfile_gammas, simtelfile_protons,
+    print("Total rate triggered gamma [Hz]", total_rate_gamma)
+
     # Weight events and count number of events per bin:
     for i in range(0,eb):  # binning in energy
-        for j in range(0,gb):  # cut in gammaness
-            for k in range(0,tb):  # cut in theta2
-                eg_w_sum = np.sum(e_reco_gw[(e_reco_g < E[i+1]) & (e_reco_g > E[i]) \
+        total_rate_proton_ebin = np.sum(rate_weighted_p[(e_reco_p < E[i+1]) & (e_reco_p > E[i])])
+        total_rate_gamma_ebin = np.sum(rate_weighted_g[(e_reco_g < E[i+1]) & (e_reco_g > E[i])])
+
+        print("********Energy bin [GeV]*****")
+        print(E[i], E[i+1])
+        print("**************")
+        print("Total rate triggered proton in this bin [Hz]", total_rate_proton_ebin)
+        print("Total rate triggered gamma in this bin [Hz]", total_rate_gamma_ebin)
+
+        for j in range(0,gb):  #  cut in gammaness
+            for k in range(0,tb):  #  cut in theta2                
+                rate_g_ebin = np.sum(rate_weighted_g[(e_reco_g < E[i+1]) & (e_reco_g > E[i]) \
                                             & (gammaness_g > g[j]) & (theta2_g < t[k])])
 
-                ep_w_sum = np.sum(e_reco_pw[(e_reco_p < E[i+1]) & (e_reco_p > E[i]) \
+                rate_p_ebin = np.sum(rate_weighted_p[(e_reco_p < E[i+1]) & (e_reco_p > E[i]) \
                                             & (gammaness_p > g[j]) & p_contained])
-                final_gamma[i][j][k] = eg_w_sum * obstime
-                final_hadrons[i][j][k] = ep_w_sum * obstime * area_ratio_p[k]
+                final_gamma[i][j][k] = rate_g_ebin * obstime
+                final_hadrons[i][j][k] = rate_p_ebin * obstime * area_ratio_p[k]
 
                 pre_gamma[i][j][k] = e_reco_g[(e_reco_g < E[i+1]) & (e_reco_g > E[i]) \
                                             & (gammaness_g > g[j]) & (theta2_g < t[k])].shape[0]
                 pre_hadrons[i][j][k] = e_reco_p[(e_reco_p < E[i+1]) & (e_reco_p > E[i]) \
                                             & (gammaness_p > g[j]) & p_contained].shape[0]
 
-                ngamma_per_ebin[i] = np.sum(e_reco_gw[(e_reco_g < E[i+1]) & (e_reco_g > E[i])]) * obstime
-                nhadron_per_ebin[i] = np.sum(e_reco_pw[(e_reco_p < E[i+1]) & (e_reco_p > E[i])]) * obstime
+                ngamma_per_ebin[i] = np.sum(rate_weighted_g[(e_reco_g < E[i+1]) & (e_reco_g > E[i])]) * obstime
+                nhadron_per_ebin[i] = np.sum(rate_weighted_p[(e_reco_p < E[i+1]) & (e_reco_p > E[i])]) * obstime
 
     nex_5sigma, sens = calculate_sensitivity_lima(final_gamma, final_hadrons * noff, 1/noff,
                                                   eb, gb, tb)
     # Avoid bins which are empty or have too few events:
-    min_num_events = 10
-    min_pre_events = 10
+    min_num_events = 5
+    min_pre_events = 5
+
     # Minimum number of gamma and proton events in a bin to be taken into account for minimization
     for i in range(0, eb):
```

Comment:
This is a very inefficient way of dealing with numpy arrays.
Suggestion:
```
conditions_good_senstivity = (sens > 0) \
                             * (final_hadrons > min_num_events) \
                             * (pre_gamma > min_pre_events) \
                             * (pre_hadrons > min_pre_events)

sens[~conditions_good_senstivity] = np.inf
```
If any of the condition is not respected at position `i,j,k`, the value `sens[i,j,k]` is affected `np.inf` (works whatever the shape of the arrays, but all arrays need to have the same shape - which should be required but is not currently enforced)